pallets/utxo/src/staking.rs

```diff
@@ -0,0 +1,280 @@
+    fn get_account_id(pubkey: &H256) -> AccountId;
+
+    /// start the staking.
+    /// # Arguments
+    /// * `stash_account` - A placeholder of the "supposed" validator. This is only to "satisfy"
+    /// the `pallet-staking`'s needs to be able to stake.
+    /// * `controller_account` - The ACTUAL validator. But this is NOT SO, in the `pallet-staking`.
+    /// In `pallet-staking`, its job is like an "accountant" to the stash account.
+    /// * `session_key` - to get up-to-date with validators, eras, sessions. see `pallet-session`.
+    /// * `value` - the amount to stake/bond/stash
+    fn stake(stash_account:&AccountId, controller_account:&AccountId, session_key:&mut Vec<u8>, value:Value) -> DispatchResultWithPostInfo;
+
+    /// stake more funds for the validator
+    fn stake_extra(controller_account:&AccountId, value:Value) -> DispatchResultWithPostInfo;
+
+    /// quitting the role of a validator.
+    fn pause(controller_account:&AccountId) -> DispatchResultWithPostInfo;
+
+    /// transfer balance from the locked state to the actual free balance.
+    fn withdraw(controller_account: &AccountId) -> DispatchResultWithPostInfo;
+
+}
+
+/// performs the staking OUTSIDE of the `pallet-utxo`. Calls the `fn stake(...)` function of
+/// `StakingHelper` trait.
+/// # Arguments
+/// * `stash_pubkey` - is a `validator` in terms of `pallet-staking`; but in the utxo, this is nothing.
+/// This is only to satisfy the `pallet-staking`. BUT this acts as a "bank", so this account MUST have
+/// funds in its `pallet-balances` counterpart.
+/// * `controller_pubkey` - is the `validator` in our utxo system.
+/// * `session_key` - to get up-to-date with the validators, eras, sessions. see `pallet-session`.
+/// * `value` - the amount to stake/bond/stash
+pub(crate) fn stake<T: Config>(
+    stash_pubkey: &H256,
+    controller_pubkey: &H256,
+    session_key:&Vec<u8>,
+    value:Value
+)
+    -> DispatchResultWithPostInfo {
+    ensure!(!<StakingCount<T>>::contains_key(controller_pubkey),Error::<T>::StakingAlreadyExists);
+
+    let mut session_key = session_key.to_vec();
+
+    let stash_account = T::StakingHelper::get_account_id(stash_pubkey);
+    let controller_account = T::StakingHelper::get_account_id(controller_pubkey);
+
+    let non_mlt_value = value.checked_div(MLT_UNIT).ok_or("could not convert to a non mlt value.")?;
+    T::StakingHelper::stake(&stash_account, &controller_account, &mut session_key,non_mlt_value)
+}
+
+/// stake more values. This is only for existing validators.
+pub(crate) fn stake_extra<T: Config>(controller_pubkey: &H256, value:Value) -> DispatchResultWithPostInfo {
+    // Checks whether a given pubkey is a validator
+    ensure!(<StakingCount<T>>::contains_key(controller_pubkey.clone()) , Error::<T>::NoStakingRecordFound);
+
+    let controller_account= T::StakingHelper::get_account_id(controller_pubkey);
+
+    let non_mlt_value = value.checked_div(MLT_UNIT).ok_or("could not convert to a non mlt value.")?;
```

Comment:
Are you converting from x * 1_000 * 100_000_000 to just x here? How will this handle decimals? Say I stake 40_000.50 MLT?

Reply:
question. in the UI, it's only **`x`**? I'm wondering whether  the js backend of the UI will convert it into `x * 1_000 * 100_000_000 t`, and the rust side will receive it in that way;

or the rust side will receive **`x`** , and will compute it to `* 1_000 * 100_000_000`?